### PR TITLE
FIX-#2074: CSVReader changes index_col if length names mismatch

### LINKED ISSUE #2074: read_csv with Ray engine works incorrectly if len(names) is less than actual number of columns in csv file
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
- **Modin version** (`modin.__version__`): 0.7.3+350.g2ca3f34
- **Python version**: 3.8.5
- **Code we can use to reproduce**:
```
import os
os.environ["MODIN_ENGINE"] = "ray"

import pandas
import modin.pandas as pd
from modin.pandas.test.utils import df_equals

test_filename = "test.csv"
csv_single_col = """one,two
three,four
"""

try :
    with open(test_filename, "w") as f:
        f.write(csv_single_col)

    df_pandas = pandas.read_csv(test_filename, names=["col2"])
    print(df_pandas)
    df_pd = pd.read_csv(test_filename, names=["col2"])
    print(df_pd)
    df_equals(df_pd, df_pandas)
finally:
    os.remove(test_filename)
```

### Source code / logs

```
       col2
one     two
three  four
   col2
0   two
1  four
Traceback (most recent call last):
  File "test_read_csv.py", line 47, in <module>
    df_equals(df_pd, df_pandas)
  File "/localdisk/amyskov/modin2/modin/pandas/test/utils.py", line 499, in df_equals
    assert_frame_equal(
  File "/localdisk/amyskov/miniconda3_2/envs/modin_test3/lib/python3.8/site-packages/pandas/_testing.py", line 1332, in assert_frame_equal
    assert_index_equal(
  File "/localdisk/amyskov/miniconda3_2/envs/modin_test3/lib/python3.8/site-packages/pandas/_testing.py", line 679, in assert_index_equal
    _testing.assert_almost_equal(
  File "pandas/_libs/testing.pyx", line 65, in pandas._libs.testing.assert_almost_equal
  File "pandas/_libs/testing.pyx", line 174, in pandas._libs.testing.assert_almost_equal
  File "/localdisk/amyskov/miniconda3_2/envs/modin_test3/lib/python3.8/site-packages/pandas/_testing.py", line 915, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.index are different

DataFrame.index values are different (100.0 %)
[left]:  RangeIndex(start=0, stop=2, step=1)
[right]: Index(['one', 'three'], dtype='object')
```

### PR DESCRIPTION
**Pull request by kvu35**
#2074   What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2074 
- [x] tests added and passing

### CODE REVIEW

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
@kvu35 Thanks for the PR, the diff doesn't look quite right. There's a 150,000 line change but I think it was unintentional. Can you take a look?

**Comment by kvu35** [comment]
@devin-petersohn Hi Devin, sorry for the the delay. I've fixed the issue in the PR.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Great @kvu35, thanks for the rebase! I think we're just missing the test that reproduces the error and shows it is fixed. This test will go in `modin/pandas/test/test_io.py`.

**Comment by kvu35** [comment]
@devin-petersohn Sorry about that, tests added, but I noticed that the omnisci backend check is no longer passing. However, the log tells me that the pytest had no failing tests. Can you help me figure out the source of the error? I don't have the proper environment to run these omnisci tests.

**Comment by Garra1980** [comment]
This is a problem on omnisci backend side, we are looking into this. Since all tests passed it can be assumed that all is clear

File: modin/pandas/test/test_io.py [hunk 1]
```diff
@@ -1707,3 +1707,19 @@ def test_from_arrow():
...
     df_equals(modin_df, pandas_df)
+
+
+@pytest.mark.parametrize(
+    "kwargs",
+    [
+        {"names": [0, 1, 2, 3, 4, 5]},
+        {"names": [0]},
+        {"names": None, "usecols": [0, 1, 2]},
+        {"names": [0, 1, 2], "usecols": [0, 1, 2]},
```
Comment (devin-petersohn) [inline modin/pandas/test/test_io.py:1718]: I know it does not matter implementation-wise, but for good measure let's test an unordered list as well.

File: modin/engines/base/io/text/csv_reader.py [hunk 2]
```diff
@@ -67,13 +68,20 @@ def _read(cls, filepath_or_buffer, **kwargs):
                 filepath_or_buffer,
                 **dict(kwargs, usecols=None, nrows=0, skipfooter=0, index_col=None),
             ).columns
+        elif index_col is None and not usecols:
+            empty_pd_df = pandas.read_csv(filepath_or_buffer, nrows=0)
+            num_cols = len(empty_pd_df.columns)
+            if num_cols > len(names):
+                index_col = [i for i in range(0, num_cols - len(names))]
```
Comment (devin-petersohn) [inline modin/engines/base/io/text/csv_reader.py:75]:
```suggestion
                index_col = list(range(num_cols - len(names)))
```

File: modin/engines/base/io/text/csv_reader.py [hunk 3]
```diff
@@ -67,13 +68,20 @@ def _read(cls, filepath_or_buffer, **kwargs):
                 filepath_or_buffer,
                 **dict(kwargs, usecols=None, nrows=0, skipfooter=0, index_col=None),
             ).columns
+        elif index_col is None and not usecols:
+            empty_pd_df = pandas.read_csv(filepath_or_buffer, nrows=0)
```
Comment (devin-petersohn) [inline modin/engines/base/io/text/csv_reader.py:72]: A brief comment here on why this is needed/what the resulting behavior is would be really helpful.
```suggestion
            # When names is set to some list that is smaller than the number of columns
            # in the file, the first columns are built as a hierarchical index.
            empty_pd_df = pandas.read_csv(filepath_or_buffer, nrows=0)
```

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
A couple of minor nit comments for future maintenance.

File: modin/engines/base/io/text/csv_reader.py (hunk 2 above)
Comment (kvu35) [reply modin/engines/base/io/text/csv_reader.py]: Done, rewrote the line as suggested.

File: modin/pandas/test/test_io.py (hunk 1 above)
Comment (kvu35) [reply modin/pandas/test/test_io.py]: Changed to use random ordering for usecols and naming.

File: modin/engines/base/io/text/csv_reader.py (hunk 3 above)
Comment (kvu35) [reply modin/engines/base/io/text/csv_reader.py:74]: Comments added.

**Review by devin-petersohn — APPROVED** [review]
Thanks @kvu35 LGTM!

**Pull request merged**